File: Homework/Objective-C/HW31-32/HW31-32/ViewController.m
//
//  ViewController.m
//  HW31-32
//
//  Created by Руслан on 12.11.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

#import "ViewController.h"
#import "RTFootballClub.h"

@interface ViewController () <UITableViewDelegate, UITableViewDataSource>

@property (weak, nonatomic) UITableView *tableView;
@property (strong, nonatomic) NSMutableArray *footballLeagues;

@end

@implementation ViewController

- (void) loadView {
    [super loadView];
    
    UITableView *tableView = [[UITableView alloc] initWithFrame:self.view.bounds
                                                          style:UITableViewStyleGrouped];
    
    tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    
    tableView.delegate = self;
    tableView.dataSource = self;
    
    [self.view addSubview:tableView];
    
    self.tableView = tableView;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.footballLeagues = [[NSMutableArray alloc] init];
    
    for (int i = 0; i < 2; i++) {
        
        NSArray *randomLeague = [RTFootballClub randomFootballLeague];
        randomLeague = [self sortedByPoints:randomLeague];
        [self.footballLeagues addObject:randomLeague];
    }
    
    self.navigationItem.title = @"English Football Leagues";
    
    UIBarButtonItem *editButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemEdit
                                                                                target:self
                                                                                action:@selector(actionEdit:)];
    self.navigationItem.rightBarButtonItem = editButton;
    
    UIBarButtonItem *newLeagueButton = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
                                                                                     target:self
                                                                                     action:@selector(actionNewLeague:)];
    self.navigationItem.leftBarButtonItem = newLeagueButton;
    
    [self.tableView reloadData];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

#pragma mark - Actions : Edit and Add

- (void) actionEdit:(UIBarButtonItem *)editButton {
    
    BOOL isEditing = self.tableView.isEditing;
    
    [self.tableView setEditing:!isEditing animated:YES];
    
    UIBarButtonSystemItem item = UIBarButtonSystemItemDone;
    
    if (isEditing) {
        item = UIBarButtonSystemItemEdit;
    }
    
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:item
                                                                                           target:self
                                                                                           action:@selector(actionEdit:)];
    
    [[UIApplication sharedApplication] beginIgnoringInteractionEvents];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        if ([[UIApplication sharedApplication] isIgnoringInteractionEvents]) {
            [[UIApplication sharedApplication] endIgnoringInteractionEvents];
        }
    });
}

- (void) actionNewLeague:(UIBarButtonItem *)newLeague {
    
    NSArray *league = [RTFootballClub makeEmptyLeague];
    [self.footballLeagues insertObject:league atIndex:0];
    
    [self.tableView beginUpdates];
    
    NSIndexSet *indexSet = [NSIndexSet indexSetWithIndex:0];
    [self.tableView insertSections:indexSet withRowAnimation:UITableViewRowAnimationAutomatic];
    
    [self.tableView endUpdates];
    
}

#pragma mark - Helpful functions

- (NSArray *) sortedByPoints:(NSArray *)array {
    
    array = [array sortedArrayUsingComparator:^NSComparisonResult(RTFootballClub *club1, RTFootballClub *club2) {
        
        return club1.points < club2.points;
        
    }];
    
    return array;
}

- (UIColor *) pointsColor:(NSInteger)points {
    
    if (points < 30) {
        return [UIColor redColor];
    } else if (points < 60) {
        return [UIColor orangeColor];
    } else if (points < 90) {
        return [UIColor magentaColor];
    } else {
        return [UIColor greenColor];
    }
}

- (UITableViewCell *) addFirstCell {
    
    UITableViewCell *firstCell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                                        reuseIdentifier:@"firstCell"];
    firstCell.textLabel.text = @"Tap to add new club";
    firstCell.textLabel.textColor = [UIColor lightGrayColor];
    firstCell.textLabel.alpha = 0.75f;
    
    return firstCell;
}

- (UITableViewCell *) addLastCell {
    
    UITableViewCell *lastCell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault
                                                        reuseIdentifier:@"lastCell"];
    lastCell.textLabel.text = @"Tap to delete this league";
    lastCell.textLabel.textColor = [UIColor lightGrayColor];
    lastCell.textLabel.alpha = 0.75f;
    
    return lastCell;
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    
    return [[self.footballLeagues objectAtIndex:section] count];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return [self.footballLeagues count];
}

- (nullable NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section {
    
    return [[[self.footballLeagues objectAtIndex:section] firstObject] league];
}

//                                                  *** DATA ROWS ***

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSInteger lastIndex = [[self.footballLeagues objectAtIndex:indexPath.section] count] - 1;
    
    if (indexPath.row == 0) {
        
        static NSString *identifier = @"firstCell";
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        
        if (!cell) {
            cell = [self addFirstCell];
        }
        
        return cell;
    } else if (indexPath.row != lastIndex) {
        
        static NSString *identifier = @"cell";
        
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        
        if (!cell) {
            cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleValue1 reuseIdentifier:identifier];
        }
        
        RTFootballClub *club = [[self.footballLeagues objectAtIndex:indexPath.section] objectAtIndex:indexPath.row];
        
        cell.textLabel.text = club.name;
        cell.detailTextLabel.text = [NSString stringWithFormat:@"%ld", club.points];
        cell.detailTextLabel.textColor = [self pointsColor:club.points];
        
        return cell;
    } else {
        
        static NSString *identifier = @"lastCell";
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
        
        if (!cell) {
            cell = [self addLastCell];
        }
        
        return cell;
        
    }
}
//                                              --------------------------

//                                                  *** MOVING ROWS ***

- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSInteger lastIndex = [[self.footballLeagues objectAtIndex:indexPath.section] count] - 1;
    
    return indexPath.row != 0 && indexPath.row != lastIndex ? YES : NO;
}

- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)sourceIndexPath
                                                  toIndexPath:(NSIndexPath *)destinationIndexPath {
    
    NSArray *sourceLeague = [self.footballLeagues objectAtIndex:sourceIndexPath.section];
    
    RTFootballClub *sourceClub = [sourceLeague objectAtIndex:sourceIndexPath.row];
    NSMutableArray *tempLeague = [[NSMutableArray alloc] initWithArray:sourceLeague];
    
    if (sourceIndexPath.section == destinationIndexPath.section) {
        
        [tempLeague removeObject:sourceClub];
        [tempLeague insertObject:sourceClub atIndex:destinationIndexPath.row];
        [self.footballLeagues setObject:tempLeague atIndexedSubscript:sourceIndexPath.section];
        
    } else {
        
        [tempLeague removeObject:sourceClub];
        [self.footballLeagues setObject:tempLeague atIndexedSubscript:sourceIndexPath.section];
        
        NSArray *destinationLeague = [self.footballLeagues objectAtIndex:destinationIndexPath.section];
        tempLeague = [[NSMutableArray alloc] initWithArray:destinationLeague];
        
        sourceClub.league = [[tempLeague firstObject] league];
        [tempLeague insertObject:sourceClub atIndex:destinationIndexPath.row];
        [self.footballLeagues setObject:tempLeague atIndexedSubscript:destinationIndexPath.section];
    }
    
    [self.tableView reloadData];
}
//                                              --------------------------

//                                                  *** DELETE ROWS ***

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle
                                            forRowAtIndexPath:(NSIndexPath *)indexPath {
    
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        
        NSArray *footballLeague = [self.footballLeagues objectAtIndex:indexPath.section];
        RTFootballClub *club = [footballLeague objectAtIndex:indexPath.row];
        
        NSMutableArray *tempFootballLeague = [[NSMutableArray alloc] initWithArray:footballLeague];
        [tempFootballLeague removeObject:club];
        [self.footballLeagues setObject:tempFootballLeague atIndexedSubscript:indexPath.section];
    }
    
    [self.tableView beginUpdates];
    
    [self.tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
    
    [self.tableView endUpdates];
}

//                                              --------------------------

#pragma mark - UITableViewDelegate

- (UITableViewCellEditingStyle)tableView:(UITableView *)tableView editingStyleForRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSInteger lastIndex = [[self.footballLeagues objectAtIndex:indexPath.section] count] - 1;
    
    return indexPath.row != 0 && indexPath.row != lastIndex ? UITableViewCellEditingStyleDelete : UITableViewCellEditingStyleNone;
}

- (BOOL)tableView:(UITableView *)tableView shouldIndentWhileEditingRowAtIndexPath:(NSIndexPath *)indexPath {
    
    return NO;
}

- (nullable NSString *)tableView:(UITableView *)tableView titleForDeleteConfirmationButtonForRowAtIndexPath:(NSIndexPath *)indexPath {
    return @"Remove";
}

//                                                  *** ADDING ROWS ***

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    NSInteger lastIndex = [[self.footballLeagues objectAtIndex:indexPath.section] count] - 1;
    
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    if (indexPath.row == 0) {
        
        __block NSArray *league = [self.footballLeagues objectAtIndex:indexPath.section];
        NSMutableArray *tempLeague = [[NSMutableArray alloc] initWithArray:league];
        
        RTFootballClub *newFootballClub = [RTFootballClub randomFootballClub];
        
        [tempLeague insertObject:newFootballClub atIndex:1]; // because at index 0 we have "default club"
        league = tempLeague;
        [self.footballLeagues setObject:league atIndexedSubscript:indexPath.section];
        
        [self.tableView beginUpdates];
        
        NSIndexPath *firstIndexPath = [NSIndexPath indexPathForRow:1
                                                         inSection:indexPath.section];
        
        [self.tableView insertRowsAtIndexPaths:@[firstIndexPath] withRowAnimation:UITableViewRowAnimationAutomatic];
        
        [self.tableView endUpdates];
        
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.7 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            
            league = [self sortedByPoints:tempLeague];
            [self.footballLeagues setObject:league atIndexedSubscript:indexPath.section];
            
            [self.tableView beginUpdates];
            
            NSIndexPath *startIndexPath = [NSIndexPath indexPathForRow:1 inSection:indexPath.section];
            
            NSIndexPath *endIndexPath = [NSIndexPath indexPathForRow:[league indexOfObject:newFootballClub]
                                                           inSection:indexPath.section];
            
            [self.tableView moveRowAtIndexPath:startIndexPath toIndexPath:endIndexPath];
            
            [self.tableView endUpdates];
            
        });
        
        [[UIApplication sharedApplication] beginIgnoringInteractionEvents];
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            if ([[UIApplication sharedApplication] isIgnoringInteractionEvents]) {
                [[UIApplication sharedApplication] endIgnoringInteractionEvents];
            }
        });
//                                              --------------------------
        
//                                                  *** DELETE SECTIONS ***
        
    } else if (indexPath.row == lastIndex) {
        
        [self.footballLeagues removeObjectAtIndex:indexPath.section];
        
        [self.tableView beginUpdates];
        
        NSIndexSet *newIndexSet = [NSIndexSet indexSetWithIndex:indexPath.section];
        
        [self.tableView deleteSections:newIndexSet withRowAnimation:UITableViewRowAnimationAutomatic];
        
        [self.tableView endUpdates];
    }
}
//                                              --------------------------

- (NSIndexPath *)tableView:(UITableView *)tableView targetIndexPathForMoveFromRowAtIndexPath:(NSIndexPath *)sourceIndexPath
       toProposedIndexPath:(NSIndexPath *)proposedDestinationIndexPath {
    
    NSInteger lastIndex = [[self.footballLeagues objectAtIndex:proposedDestinationIndexPath.section] count] - 1;
    
    if (proposedDestinationIndexPath.row == 0 ||
        proposedDestinationIndexPath.row == lastIndex) {
        return sourceIndexPath;
    } else {
        return proposedDestinationIndexPath;
    }
}

@end

File: Homework/Swift/HW31-32/HW31-32/ViewController.swift
//
//  ViewController.swift
//  HW31-32
//
//  Created by Руслан on 13.11.16.
//  Copyright © 2016 Ruslan Timchenko. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

// MARK: - Immutable identifiers
    
    let cellIdentifier = "cell"
    let firstCellIdentifier = "firstCell"
    let lastCellIdentifier = "lastCell"
    
// MARK: - Mutable variables
    
    var tableView = UITableView()
    var footballLeagues = [[FootballClub]]()
    
// MARK: - Override functions
    
    override func loadView() {
        super.loadView()
        
        let tempTableView = UITableView(frame: self.view.bounds, style: .plain)
        
        tempTableView.autoresizingMask = [.flexibleHeight, .flexibleWidth]
        
        tempTableView.delegate = self
        tempTableView.dataSource = self
        
        self.view.addSubview(tempTableView)
        
        self.tableView = tempTableView
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        
        for _ in 0..<2 {
            
            var randomLeague = FootballClub.randomFootballLeague()
            sortedByPoints(&randomLeague)
            footballLeagues.append(randomLeague)
        }
        
        self.navigationItem.title = "English Football Leagues"
        
        let editButton = UIBarButtonItem(barButtonSystemItem: .edit, target: self, action: #selector(actionEdit))
        self.navigationItem.rightBarButtonItem = editButton
        
        let addButton = UIBarButtonItem(barButtonSystemItem: .add, target: self, action: #selector(actionAdd))
        self.navigationItem.leftBarButtonItem = addButton
        
        self.tableView.reloadData()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
// MARK: - Actions: Edit and Add
    
    func actionEdit(_ sender: UIBarButtonItem) {
        
        let isEditing = self.tableView.isEditing
        
        self.tableView.setEditing(!isEditing, animated: true)
        
        var item = UIBarButtonSystemItem.done
        
        if isEditing {
            item = UIBarButtonSystemItem.edit
        }
        
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(barButtonSystemItem: item,
                                                                 target: self,
                                                                 action: #selector(actionEdit))
        
        prohibitAnyActionsDuringAnimation(with: 0.3)
    }
    
    func actionAdd(_ sender: UIBarButtonItem) {
        
        let league = FootballClub.makeEmptyLeague()
        footballLeagues.insert(league, at: 0)
        
        self.tableView.beginUpdates()
        
        let indexSet = IndexSet.init(integer: 0)
        self.tableView.insertSections(indexSet, with: .automatic)
        
        self.tableView.endUpdates()
        
        prohibitAnyActionsDuringAnimation(with: 0.3)
        
    }
    
// MARK: - Helpful functions
    
    func prohibitAnyActionsDuringAnimation(with time: Double) {
        
        UIApplication.shared.beginIgnoringInteractionEvents()
        DispatchQueue.main.asyncAfter(deadline: .now() + time) {
            
            if UIApplication.shared.isIgnoringInteractionEvents {
                UIApplication.shared.endIgnoringInteractionEvents()
            }
            
        }
    }
    
    func sortedByPoints(_ array: inout [FootballClub]) {
        
        array = array.sorted { club1, club2 in
            return club1.points > club2.points
        }
    }
    
    func pointsColor(_ points: Int) -> UIColor {
        
        switch points {
        case 0..<30:
            return UIColor.red
        case 0..<60:
            return UIColor.orange
        case 0..<90:
            return UIColor.magenta
        default:
            return UIColor.green
        }
    }
    
    func addFirstCell() -> UITableViewCell {
        
        let firstCell = UITableViewCell(style: .value1, reuseIdentifier: firstCellIdentifier)
        
        firstCell.textLabel?.text = "Tap to add new club"
        firstCell.textLabel?.textColor = UIColor.gray
        firstCell.textLabel?.alpha = 0.85
        
        return firstCell
    }
    
    func addLastCell() -> UITableViewCell {
        
        let lastCell = UITableViewCell(style: .value1, reuseIdentifier: lastCellIdentifier)
        
        lastCell.textLabel?.text = "Tap to delete this league"
        lastCell.textLabel?.textColor = UIColor.gray
        lastCell.textLabel?.alpha = 0.85
        
        return lastCell
    }

// MARK: - UITableViewDataSource
    
    public func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return footballLeagues[section].count
    }
    
    public func numberOfSections(in tableView: UITableView) -> Int {
        
        return footballLeagues.count
    }
    
    public func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        
        return footballLeagues[section].first?.league
    }
    
//                                                          *** DATA ROWS ***
    
    public func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        switch indexPath.row {
        case 0:
            
            var firstCell = tableView.dequeueReusableCell(withIdentifier: firstCellIdentifier)
            
            if firstCell == nil {
                firstCell = addFirstCell()
            }
            return firstCell!
            
        case self.footballLeagues[indexPath.section].endIndex - 1:
            
            var lastCell = tableView.dequeueReusableCell(withIdentifier: lastCellIdentifier)
            
            if lastCell == nil {
                lastCell = addLastCell()
            }
            return lastCell!
            
        default:
            
            var cell = tableView.dequeueReusableCell(withIdentifier: cellIdentifier)
            
            if cell == nil {
                cell = UITableViewCell(style: .value1, reuseIdentifier: cellIdentifier)
            }
            
            let club = footballLeagues[indexPath.section][indexPath.row]
            
            cell?.textLabel?.text = club.name
            cell?.detailTextLabel?.text = "\(club.points)"
            cell?.detailTextLabel?.textColor = pointsColor(club.points)
            
            return cell!
        }
    }
//                                                   -----------------------------
    
//                                                          *** MOVING ROWS ***

    public func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        
        return indexPath.row != 0 && indexPath.row != footballLeagues[indexPath.section].endIndex - 1
    }
    
    public func tableView(_ tableView: UITableView, moveRowAt sourceIndexPath: IndexPath, to destinationIndexPath: IndexPath) {
        
        var tempLeague = footballLeagues[sourceIndexPath.section]
        let sourceClub = tempLeague[sourceIndexPath.row]
        
        if sourceIndexPath.section == destinationIndexPath.section {
            
            tempLeague.remove(at: sourceIndexPath.row)
            tempLeague.insert(sourceClub, at: destinationIndexPath.row)
            footballLeagues[sourceIndexPath.section] = tempLeague
            
        } else {
            
            tempLeague.remove(at: sourceIndexPath.row)
            footballLeagues[sourceIndexPath.section] = tempLeague
            
            tempLeague = footballLeagues[destinationIndexPath.section]
            sourceClub.league = (tempLeague.first?.league)!
            tempLeague.insert(sourceClub, at: destinationIndexPath.row)
            footballLeagues[destinationIndexPath.section] = tempLeague
        }
        
        self.tableView.reloadData()
    }
//                                                   -----------------------------
    
//                                                          *** DELETE ROWS ***
    public func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle,
                          forRowAt indexPath: IndexPath) {
        
        if editingStyle == .delete {
            
            var footballLeague = footballLeagues[indexPath.section]
            
            footballLeague.remove(at: indexPath.row)
            footballLeagues[indexPath.section] = footballLeague
        }
        
        self.tableView.beginUpdates()
        
        self.tableView.deleteRows(at: [indexPath], with: .automatic)
        
        self.tableView.endUpdates()
    }
//                                                   -----------------------------
    
// MARK: - UITableViewDelegate
    
    public func tableView(_ tableView: UITableView,
                                   editingStyleForRowAt indexPath: IndexPath) -> UITableViewCellEditingStyle {
        
        return (indexPath.row != 0 && indexPath.row != footballLeagues[indexPath.section].endIndex - 1) ? .delete : .none
    }
    
    public func tableView(_ tableView: UITableView, shouldIndentWhileEditingRowAt indexPath: IndexPath) -> Bool {
        return false
    }
    
    public func tableView(_ tableView: UITableView, titleForDeleteConfirmationButtonForRowAt indexPath: IndexPath) -> String? {
        return "Remove"
    }
    
    public func tableView(_ tableView: UITableView, targetIndexPathForMoveFromRowAt
                sourceIndexPath: IndexPath, toProposedIndexPath proposedDestinationIndexPath: IndexPath) -> IndexPath {
        
        if proposedDestinationIndexPath.row == 0 ||
           proposedDestinationIndexPath.row == footballLeagues[proposedDestinationIndexPath.section].endIndex - 1 {
            return sourceIndexPath
        } else {
            return proposedDestinationIndexPath
        }
    }
//                                                          *** ADDING ROWS ***
    public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        self.tableView.deselectRow(at: indexPath, animated: true)
        
        if indexPath.row == 0 {
            
            var tempLeague = footballLeagues[indexPath.section]
            let newFootballClub = FootballClub.randomFootballClub()
            
            tempLeague.insert(newFootballClub, at: 1)
            footballLeagues[indexPath.section] = tempLeague
            
            self.tableView.beginUpdates()
            let startIndexPath = IndexPath(row: 1, section: indexPath.section)
            self.tableView.insertRows(at: [startIndexPath], with: .automatic)
            self.tableView.endUpdates()
            
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.7, execute: { 
                
                self.sortedByPoints(&tempLeague)
                self.footballLeagues[indexPath.section] = tempLeague
                
                self.tableView.beginUpdates()
                let endIndexPath = IndexPath(row: tempLeague.index(of: newFootballClub)!, section: indexPath.section)
                self.tableView.moveRow(at: startIndexPath, to: endIndexPath)
                self.tableView.endUpdates()
    
            })
            self.prohibitAnyActionsDuringAnimation(with: 1.0)
//                                                   -----------------------------
            
//                                                          *** DELETE SECTIONS ***
        } else if indexPath.row == footballLeagues[indexPath.section].endIndex - 1 {
            
            footballLeagues.remove(at: indexPath.section)
            
            self.tableView.beginUpdates()
            let newIndexSet = IndexSet.init(integer: indexPath.section)
            self.tableView.deleteSections(newIndexSet, with: .automatic)
            self.tableView.endUpdates()
            
        }
        
    }

}
